Branch-SDK/src/main/java/io/branch/referral/BillingGooglePlay.kt

```diff
@@ -0,0 +1,273 @@
+                    currency = buo.contentMetadata.currencyType
+                }
+
+                if (contentItemBUOs.isNotEmpty()) {
+                    createAndLogEventForPurchase(
+                        context,
+                        purchase,
+                        contentItemBUOs,
+                        currency,
+                        revenue,
+                        "In-App Purchase"
+                    )
+                }
+            } else {
+                PrefHelper.LogException("Failed to query subscriptions. Error: " + billingResult.debugMessage, Exception())
+            }
+        }
+    }
+
+    private fun createBUOWithSubsProductDetails(product: ProductDetails?): BranchUniversalObject {
+        if (product != null) {
+
+            val pricingPhaseList =
+                product.subscriptionOfferDetails?.get(0)?.pricingPhases?.pricingPhaseList?.get(0)
+
+            val currency = pricingPhaseList?.let {
+                CurrencyType.valueOf(
+                    it.priceCurrencyCode
+                )
+            }
+
+            val price = pricingPhaseList?.priceAmountMicros?.div(1000000.0)
+
+            if (price != null && currency != null) {
+
+                return BranchUniversalObject()
+                    .setCanonicalIdentifier(product.productId)
+                    .setTitle(product.title)
+                    .setContentMetadata(
+                        ContentMetadata()
+                            .addCustomMetadata("product_type", product.productType)
+                            .setPrice(price, currency)
+                            .setProductName(product.name)
+                            .setQuantity(1.0)
+                            .setContentSchema(BranchContentSchema.COMMERCE_PRODUCT)
+
+                    )
+            } else {
+                return BranchUniversalObject()
+                    .setCanonicalIdentifier(product.productId)
+                    .setTitle(product.title)
+                    .setContentMetadata(
+                        ContentMetadata()
+                            .addCustomMetadata("product_type", product.productType)
+                            .setProductName(product.name)
+                            .setQuantity(1.0)
+                            .setContentSchema(BranchContentSchema.COMMERCE_PRODUCT)
+                    )
+            }
+        } else {
+            return BranchUniversalObject()
+        }
+    }
+
+    private fun createBUOWithInAppProductDetails(
+        product: ProductDetails?,
+        quantity: Int
+    ): BranchUniversalObject {
+        if (product != null) {
+
+            val currency = product.oneTimePurchaseOfferDetails?.priceCurrencyCode?.let {
+                CurrencyType.valueOf(it)
+            }
+            val price = product.oneTimePurchaseOfferDetails?.priceAmountMicros?.div(1000000.0)
+
+            if (price != null && currency != null) {
```

Comment:
Because the only thing that changes between these two large if blocks is setting the price, we can reduce this code to:

```
val buo = BranchUniversalObject()
                    .setCanonicalIdentifier(product.productId)
                    .setTitle(product.title))
val contentMetadata = ContentMetadata()
                            .addCustomMetadata("product_type", product.productType)
                            .setProductName(product.name)
                            .setQuantity(quantity.toDouble())
                            .setContentSchema(BranchContentSchema.COMMERCE_PRODUCT)

if(price != null && currency != null){
    contentMetadata.setPrice(price, currency);
}

buo.setContentMetadat(contentMetadata);

return buo;

```